Persistent session storage

Zsailer opened the pull request:
Fixes #611

Makes the session database path a configurable trait and avoids overwriting the database if it exists.

Also, sets the database's isolation level to `None` so that it autocommits all changes after every change.

kevin-bates commented on jupyter_server/services/sessions/sessionmanager.py:
```diff
@@ -13,13 +13,19 @@
 
 class SessionManager(LoggingConfigurable):
 
+    database_path = Unicode(
```
Since this doesn't work if pointing at a directory, I suggest it be named `database_filepath` and that we perhaps add a valiator to ensure the value is not a directory.  Perhaps adding an example to the help string like "/path/to/my/dbfile.db" would also clarify the expected value.

It might be good to reference the meaning of ":memory:" in the help string as well.

FWIW, there's the log entry if given a directory:
```
[E 2021-11-19 14:44:09.911 ServerApp] Uncaught exception GET /api/sessions?1637361849903 (::1)
    HTTPServerRequest(protocol='http', host='localhost:8888', method='GET', uri='/api/sessions?1637361849903', version='HTTP/1.1', remote_ip='::1')
    Traceback (most recent call last):
      File "/opt/anaconda3/envs/elyra-dev/lib/python3.8/site-packages/tornado/web.py", line 1704, in _execute
        result = await result
      File "/opt/anaconda3/envs/elyra-dev/lib/python3.8/site-packages/jupyter_server/services/sessions/handlers.py", line 26, in get
        sessions = await ensure_async(sm.list_sessions())
      File "/opt/anaconda3/envs/elyra-dev/lib/python3.8/site-packages/jupyter_server/utils.py", line 189, in ensure_async
        result = await obj
      File "/opt/anaconda3/envs/elyra-dev/lib/python3.8/site-packages/jupyter_server/services/sessions/sessionmanager.py", line 266, in list_sessions
        c = self.cursor.execute("SELECT * FROM session")
      File "/opt/anaconda3/envs/elyra-dev/lib/python3.8/site-packages/jupyter_server/services/sessions/sessionmanager.py", line 46, in cursor
        self._cursor = self.connection.cursor()
      File "/opt/anaconda3/envs/elyra-dev/lib/python3.8/site-packages/jupyter_server/services/sessions/sessionmanager.py", line 58, in connection
        self._connection = sqlite3.connect(self.database_path, isolation_level=None)
    sqlite3.OperationalError: unable to open database file
```

mwakaba2 replied:
I agree, I think validating that the path isn't a directory would be helpful. 
If we're expecting the database to already exist before calling `connect`, I also think it'll be helpful to check if the file is a sqlite db because `sqlite3` will automatically create a new database during `connect()` if it doesn't exist.

Wh1isper replied:
Agree, just add a valid function or provide a default filename if path is a directory would be helpful.

Zsailer replied:
Thanks, all!

Renaming to `database_filepath` is a good idea. This should be more specific. I'll also add some documentation describing the meaning of `:memory:`

I'll work on adding validation per @mwakaba2's and @Wh1isper's comments too.

kevin-bates reviewed (commented):
Thanks for creating this PR Zach.  This seems like a good evolutionary step in the server.

Zsailer commented:
Ok, I've added validator to `database_filepath` trait in SessionManager that verifies:
- path is not a directory
- file on disk is a valid SQLite 3 file (empty files are also allowed).

I've added some unit tests to (1) confirm  `database_filepath` validation is working properly and (2) the session persistence is working when across session manager instances.

Ping @kevin-bates, @mwakaba2 and @Wh1isper for a second round of reviews. Thanks, y'all!

blink1073 commented on jupyter_server/services/sessions/sessionmanager.py:
```diff
@@ -13,13 +15,48 @@
+                    "`database_filepath` expected a file path, but the given path is a directory."
+                )
+            # If the file exists, but it's empty, its a valid entry.
+            if os.stat(path).st_size == 0:
```
I'd say just try to connect to the file here and let the error propagate.  I verified that you can connect to an empty file.

blink1073 commented on jupyter_server/services/sessions/sessionmanager.py:
```diff
@@ -47,14 +46,12 @@ def _validate_database_filepath(self, proposal):
-            if os.stat(path).st_size == 0:
-                return value
             # Verify that database path is an SQLite 3 Database by checking its header.
             with open(value, "rb") as f:
```
I meant also for the contents validation piece, rather than inspecting the file

Zsailer replied:
I think `sqlite3` overwrites the file if it's not a valid database format, rather than raising an exception if the file has contents that are not sqlite data.

blink1073 replied:
Ah, yup, confirmed.

blink1073 reviewed (approved):
LGTM, thanks!

Zsailer commented:
Thank you! Merging!
